src/tf/tf.cpp

```diff
@@ -0,0 +1,775 @@
+            return prog.add_instruction(x, args);
+        });
+    }
+
+    instruction_ref
+    parse_batchnorm(const std::string&, attribute_map attributes, std::vector<instruction_ref> args)
+    {
+        float epsilon                                     = 1e-5f;
+        float momentum                                    = 0.9f;
+        op::batch_norm_inference::bn_infer_mode_t bn_mode = op::batch_norm_inference::spatial;
+        if(contains(attributes, "epsilon"))
+        {
+            epsilon = attributes.at("epsilon").f();
+        }
+        op::batch_norm_inference op{epsilon, momentum, bn_mode};
+        return prog.add_instruction(op, std::move(args));
+    }
+
+    instruction_ref
+    parse_biasadd(const std::string&, const attribute_map&, std::vector<instruction_ref> args)
+    {
+        uint64_t axis = 1; // assume output of previous layer is in NCHW (broadcast on channel)
+        auto l0       = prog.add_instruction(op::broadcast{axis, args[0]->get_shape()}, args[1]);
+        return prog.add_instruction(op::add{}, args[0], l0);
+    }
+
+    instruction_ref
+    parse_concat(const std::string&, attribute_map attributes, std::vector<instruction_ref> args)
+    {
+        // get index for axis within args
+        size_t axis_idx = attributes.at("N").i();
+        size_t axis     = args[axis_idx]->eval().at<int64_t>();
+        parse_axis(axis);
+        op::concat op{axis};
+        // return only first N arguments (assuming last index is the axis value)
+        return prog.add_instruction(
+            op, std::vector<instruction_ref>(args.begin(), args.begin() + args.size() - 1));
+    }
+
+    instruction_ref parse_constant(const std::string&,
+                                   attribute_map attributes,
+                                   const std::vector<instruction_ref>&)
+    {
+        literal v       = parse_tensor(attributes.at("value").tensor());
+        auto l0         = prog.add_literal(v);
+        size_t num_axes = l0->get_shape().lens().size();
+        if(num_axes >= 4)
+        {
+            std::vector<int64_t> transpose_axes = get_axes(num_axes);
+            reorder_data(transpose_axes);
+            l0 = prog.add_instruction(op::transpose{transpose_axes}, l0);
+        }
+        return l0;
+    }
+
+    instruction_ref
+    parse_conv(const std::string&, attribute_map attributes, std::vector<instruction_ref> args)
+    {
+        op::convolution op;
+        if(contains(attributes, "padding"))
+        {
+            const std::string& pad_mode = attributes.at("padding").s();
+            if(pad_mode.find("SAME") != std::string::npos)
+            {
+                op.padding_mode = op::padding_mode_t::same;
+            }
+            else if(pad_mode.find("EXPLICIT") != std::string::npos)
+            {
+                std::vector<size_t> padding;
+                copy(attributes.at("explicit_paddings").list().i(), std::back_inserter(padding));
+                reorder_data(padding);
```

Comment:
Is this necessary? Padding is not in a nchw format. Its in the format of something like `H1, H2, W1, W2`.